### PR TITLE
CLN GH22873 Replace base excepts in pandas/core

### LINKED ISSUE #22873: Replace bare excepts by explicit excepts in pandas/core/
(superseeds #18419, same task for different files: #22872, #22874, #22875, #22877)

In several parts of the code, we have bare excepts in the form of:
```python
try:
    my_value = my_dict[my_key]
except:  # we are capturing all the exceptions, when we want to capture only KeyError
    my_value = None
```

This is a bad practice, and we want to avoid having this in the code. What we want instead is:
```python
try:
    my_value = my_dict[my_key]
except KeyError:
    my_value = None
```
Of course in every case, the exception can be different (not always `KeyError`), and some research is needed to see which is the right exception (or exceptions, it can be a tuple) that needs to be captured. In some cases, every exception needs to be captured and we'll use `except Exception:`, but this should be avoided unless we really need to capture all exceptions.

This issue is to fix all the bare excepts in testing files `pandas/tests/`. At the moment they are (note that the list can change as code evolves):
```
pandas/core/frame.py:3263:13: E722 do not use bare except'
pandas/core/frame.py:7706:9: E722 do not use bare except'
pandas/core/indexing.py:2149:9: E722 do not use bare except'
pandas/core/indexing.py:2708:13: E722 do not use bare except'
pandas/core/indexing.py:2714:9: E722 do not use bare except'
pandas/core/nanops.py:506:13: E722 do not use bare except'
pandas/core/nanops.py:818:13: E722 do not use bare except'
pandas/core/ops.py:1548:17: E722 do not use bare except'
pandas/core/window.py:2507:5: E722 do not use bare except'
pandas/core/computation/pytables.py:414:9: E722 do not use bare except'
pandas/core/dtypes/common.py:470:5: E722 do not use bare except'
pandas/core/dtypes/dtypes.py:361:9: E722 do not use bare except'
pandas/core/dtypes/dtypes.py:522:13: E722 do not use bare except'
pandas/core/indexes/frozen.py:142:9: E722 do not use bare except'
pandas/core/indexes/multi.py:1005:17: E722 do not use bare except'
pandas/core/indexes/multi.py:1010:17: E722 do not use bare except'
pandas/core/indexes/multi.py:1689:9: E722 do not use bare except'
pandas/core/indexes/multi.py:2318:17: E722 do not use bare except'
pandas/core/indexes/multi.py:2821:17: E722 do not use bare except'
pandas/core/internals/blocks.py:669:9: E722 do not use bare except'
pandas/core/internals/blocks.py:1145:9: E722 do not use bare except'
pandas/core/internals/blocks.py:1160:9: E722 do not use bare except'
pandas/core/internals/blocks.py:2441:13: E722 do not use bare except'
pandas/core/internals/blocks.py:3175:9: E722 do not use bare except'
pandas/core/sparse/array.py:309:9: E722 do not use bare except'
pandas/core/tools/datetimes.py:244:17: E722 do not use bare except'
pandas/core/tools/datetimes.py:334:9: E722 do not use bare except'
pandas/core/tools/datetimes.py:731:5: E722 do not use bare except'
pandas/core/tools/datetimes.py:738:5: E722 do not use bare except'
pandas/core/tools/datetimes.py:745:5: E722 do not use bare except'
```
An updated list can be obtained by running: `flake8 --select=E722 --config=none pandas/core/`

### PR DESCRIPTION
**Pull request by pambot**
This cleaning fix adds specific exception classes to bare except statements to correct PEP8 linting errors. The classes were chosen by a mixture of educated guessing (i.e. if a statement gets the `length` property, it might raise an `AttributeError`), trying out functions on known edge cases (i.e. giving a datetime input that is too large gives `tslibs.OutOfBoundsDatetime`), and inspecting functions in the `try` block to see what kind of errors they raise.

- [x] xref #22873
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.24.0.txt [hunk 1]
```diff
@@ -829,4 +829,5 @@ Other
 - :meth:`DataFrame.nlargest` and :meth:`DataFrame.nsmallest` now returns the correct n values when keep != 'all' also when tied on the first columns (:issue:`22752`)
 - :meth:`~pandas.io.formats.style.Styler.bar` now also supports tablewise application (in addition to rowwise and columnwise) with ``axis=None`` and setting clipping range with ``vmin`` and ``vmax`` (:issue:`21548` and :issue:`21526`). ``NaN`` values are also handled properly.
 - Logical operations ``&, |, ^`` between :class:`Series` and :class:`Index` will no longer raise ``ValueError`` (:issue:`22092`)
--
+- Added specific exception classes to fix bare except statements in `pandas/core` (:issue:``22873)
```
Comment (datapythonista) [inline doc/source/whatsnew/v0.24.0.txt]: Thanks for adding this, but not sure if it's worth adding this change to the release note. It's more of a internal change that should not have any impact on users.

File: pandas/core/indexes/multi.py [hunk 2]
```diff
@@ -975,7 +975,7 @@ def _try_mi(k):
         except KeyError as e1:
             try:
                 return _try_mi(key)
-            except KeyError:
+            except (KeyError, IndexError, ValueError):
```
Comment (datapythonista) [inline pandas/core/indexes/multi.py]: Can you clarify this change? It feels like we just wanted to ignore `KeyError` exceptions, and I'm sure you have a reason to also ignore `IndexError` and `ValueError`, but it's not obvious to me.

File: pandas/core/nanops.py [hunk 3]
```diff
@@ -815,8 +816,9 @@ def _ensure_numeric(x):
         elif is_object_dtype(x):
             try:
                 x = x.astype(np.complex128)
-            except:
-                x = x.astype(np.float64)
+            except (TypeError, ValueError):
+                raise TypeError('Could not convert {value!s} to numeric'
+                                .format(value=x))
```
Comment (datapythonista) [inline pandas/core/nanops.py]: I don't understand this change either. Seems like the original behavior was to try to cast to float. Why to not try it anymore and raise directly?

**Review by datapythonista — COMMENTED** [review]
The PR looks just great. There are couple of cases that I'm unsure, and would like to know your reasons for a change. Other than that, looks really good. Thanks for the hard work on this!

**Comment by datapythonista** [comment]
If I'm not wrong, those were problems with the build unrelated to your PR. I think the datetime one is now fix in master, and should be ok if you merge the last master. The parquet one I think it was because of an amazon url, not sure if that's fixed, or it happens just some times. We need to take a look at it, don't worry much about it, if your changes are all right, we'll fix it eventually and merge your PR.

File: pandas/core/frame.py [hunk 4]
```diff
@@ -3260,7 +3260,8 @@ def _ensure_valid_index(self, value):
         if not len(self.index) and is_list_like(value):
             try:
                 value = Series(value)
-            except:
+            except (ValueError, NotImplementedError,
+                    AssertionError, TypeError):
```
Comment (jreback) [inline pandas/core/frame.py]: where is this assertion error coming from?

File: pandas/core/indexes/multi.py [hunk 5]
```diff
@@ -1002,12 +1002,12 @@ def _try_mi(k):
                     return _try_mi(key)
                 except (KeyError):
                     raise
-                except:
+                except Exception:
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: can we be more specific here?

File: pandas/core/indexes/multi.py [hunk 6]
```diff
@@ -1002,12 +1002,12 @@ def _try_mi(k):
...
                     pass
 
                 try:
                     return _try_mi(Timestamp(key))
-                except:
+                except Exception:
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: same here

File: pandas/core/indexes/multi.py [hunk 7]
```diff
@@ -2818,7 +2818,7 @@ def _convert_can_do_setop(self, other):
                 msg = 'other must be a MultiIndex or a list of tuples'
                 try:
                     other = MultiIndex.from_tuples(other)
-                except:
+                except Exception:
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: can be more specific here

File: pandas/core/internals/blocks.py [hunk 8]
```diff
@@ -666,7 +666,7 @@ def _astype(self, dtype, copy=False, errors='raise', values=None,
 
             newb = make_block(values, placement=self.mgr_locs,
                               klass=klass, ndim=self.ndim)
-        except:
+        except Exception:
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: can be more specific here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 4 above)
Comment (pambot) [reply pandas/core/frame.py]: Maybe we shouldn't be catching this, but since it was originally supposed to catch all the things, I thought it might have been designed to catch this in `Series.__init__()`:

```            elif isinstance(data, SingleBlockManager):
                if index is None:
                    index = data.index
                elif not data.index.equals(index) or copy:
                    # GH#19275 SingleBlockManager input should only be called
                    # internally
                    raise AssertionError('Cannot pass both SingleBlockManager '
                                         '`data` argument and a different '
                                         '`index` argument.  `copy` must '
                                         'be False.')
```

File: pandas/core/internals/blocks.py (hunk 8 above)
Comment (pambot) [reply pandas/core/internals/blocks.py]: I could be wrong here as well, but the reason why I put `Exception` here is because of the context, which was to enable the following behaviour:

```        
except Exception:
        if errors == 'raise':
                raise
```

When I run the tests without this block, nothing gets raised, and I think if I try to guess which ones could get raised and other ones get raised during a user's runtime, then only the ones I guessed would get silenced by `errors` not being set to `raise`.

File: pandas/core/indexes/multi.py (hunk 5 above)
Comment (pambot) [reply pandas/core/indexes/multi.py]: Though if you tell me I should just put something anyway, I might put in `(IndexError, ValueError)` because of the slicing and the `droplevel`

File: pandas/core/frame.py (hunk 4 above)
Comment (jreback) [reply pandas/core/frame.py]: hmm, does removing AssertionError cause any tests to fail?

File: pandas/core/internals/blocks.py (hunk 8 above)
Comment (jreback) [reply pandas/core/internals/blocks.py]: ok add a ``# noqa: E722``

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by datapythonista — APPROVED** [review]
lgtm, and the problems with the build are fixed.

Thanks for taking care of this @pambot

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @pambot very nice!

**Comment by TomAugspurger** [comment]
This seems to be the first commit with the failure on numpydev. https://travis-ci.org/pandas-dev/pandas/jobs/435916255#L2767

<details>

```
____________ TestSelection.test_groupby_duplicated_column_errormsg _____________
[gw0] linux -- Python 3.7.0 /home/travis/miniconda3/envs/pandas/bin/python
self = <pandas.tests.groupby.test_grouping.TestSelection object at 0x7f87673fb668>
    def test_groupby_duplicated_column_errormsg(self):
        # GH7511
        df = DataFrame(columns=['A', 'B', 'A', 'C'],
                       data=[range(4), range(2, 6), range(0, 8, 2)])
    
>       pytest.raises(ValueError, df.groupby, 'A')
pandas/tests/groupby/test_grouping.py:42: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/core/generic.py:7136: in groupby
    observed=observed, **kwargs)
pandas/core/groupby/groupby.py:1966: in groupby
    return klass(obj, by, **kwds)
pandas/core/groupby/groupby.py:365: in __init__
    mutated=self.mutated)
pandas/core/groupby/grouper.py:588: in _get_grouper
    if is_categorical_dtype(gpr) and len(gpr) != obj.shape[axis]:
pandas/core/dtypes/common.py:574: in is_categorical_dtype
    return CategoricalDtype.is_dtype(arr_or_dtype)
pandas/core/dtypes/base.py:93: in is_dtype
    return cls.construct_from_string(dtype) is not None
pandas/core/dtypes/dtypes.py:359: in construct_from_string
    if string == 'category':
pandas/core/ops.py:1885: in f
    try_cast=False)
pandas/core/frame.py:4927: in _combine_const
    return ops.dispatch_to_series(self, other, func)
pandas/core/ops.py:1678: in dispatch_to_series
    new_data = expressions.evaluate(column_op, str_rep, left, right)
pandas/core/computation/expressions.py:205: in evaluate
    return _evaluate(op, op_str, a, b, **eval_kwargs)
pandas/core/computation/expressions.py:65: in _evaluate_standard
    return op(a, b)
pandas/core/ops.py:1658: in column_op
    for i in range(len(a.columns))}
pandas/core/ops.py:1658: in <dictcomp>
    for i in range(len(a.columns))}
pandas/core/ops.py:1510: in wrapper
    res = na_op(values, other)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
x = array([0, 2, 0]), y = 'category'
    def na_op(x, y):
        # TODO:
        # should have guarantess on what x, y can be type-wise
        # Extension Dtypes are not called here
    
        # Checking that cases that were once handled here are no longer
        # reachable.
        assert not (is_categorical_dtype(y) and not is_scalar(y))
    
        if is_object_dtype(x.dtype):
            result = _comp_method_OBJECT_ARRAY(op, x, y)
    
        elif is_datetimelike_v_numeric(x, y):
            return invalid_comparison(x, y, op)
    
        else:
    
            # we want to compare like types
            # we only want to convert to integer like if
            # we are not NotImplemented, otherwise
            # we would allow datetime64 (but viewed as i8) against
            # integer comparisons
    
            # we have a datetime/timedelta and may need to convert
            assert not needs_i8_conversion(x)
            mask = None
            if not is_scalar(y) and needs_i8_conversion(y):
                mask = isna(x) | isna(y)
                y = y.view('i8')
                x = x.view('i8')
    
            method = getattr(x, op_name, None)
            if method is not None:
                with np.errstate(all='ignore'):
>                   result = method(y)
E                   FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
```

</details>

File: pandas/core/dtypes/dtypes.py [hunk 9]
```diff
@@ -358,11 +358,11 @@ def construct_from_string(cls, string):
         try:
             if string == 'category':
                 return cls()
-        except:
+            else:
```
Comment (TomAugspurger) [inline pandas/core/dtypes/dtypes.py]: I think this *may* have been it. For whatever reason, we're here with

```
ipdb> pp string
   A  A
0  0  2
1  2  4
2  0  4
```

We do

```
ipdb> string == 'category'
*** FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
```

That FutureWarning is elevated to an error. I didn't realize this, but you can catch warnings, so the very simple fix is to also catch `FutureWarning` here, but that feels fragile. I'll investigate a better fix...

**Comment by TomAugspurger** [comment]
Moving further discussion to https://github.com/pandas-dev/pandas/issues/22960
